### PR TITLE
[SofaGeneralAnimationLoop] Improve MechanicalMatrixMapper

### PR DESCRIPTION
**Pull request by olivier-goury**
This PR adds a few boolean Data parameters for the user to have ore control when using the component:
- stopAtNodeToParse: to avoid considering children nodes of the node to be parsed.
- skipJ1tKJ1: to avoid computing the contribution of the first object (useful when several MMMappers are used and you want to avoid to compute several time the same contribution)
- skipJ2tKJ2: to avoid computing the contribution of the 2nd object

It adds also a few templates and corrects a bug (mat12.matrix->add instead of mat22.matrix->add)

Enjoy!

______________________________________________________

This PR: 
- [x] builds with SUCCESS for all platforms on the CI.
- [x] does not generate new warnings.
- [x] does not generate new unit test failures.
- [x] does not generate new scene test failures.
- [x] does not break API compatibility.
- [x] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp [hunk 1]
```diff
@@ -40,18 +40,24 @@ using namespace sofa::defaulttype;
 
 ////////////////////////////////////////////    FACTORY    //////////////////////////////////////////////
 int MechanicalMatrixMapperClass = core::RegisterObject("This component allows to map the stiffness (and mass) matrix through a mapping.")
+
+        .add< MechanicalMatrixMapper<Rigid3dTypes, Rigid3dTypes> >(true)
```
Comment (damienmarchal) [inline modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp]: I think we shouldn't anymore use explicit type in component instantiation. 
So only Rigid3Types.

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp [hunk 2]
```diff
@@ -40,18 +40,24 @@ using namespace sofa::defaulttype;
...
         .add< MechanicalMatrixMapper<Vec3Types, Rigid3Types> >(true)
         .add< MechanicalMatrixMapper<Vec3Types, Vec3Types> >(true)
         .add< MechanicalMatrixMapper<Vec1Types, Rigid3Types> >(true)
         .add< MechanicalMatrixMapper<Vec1Types, Vec1Types> >(true)
+        .add< MechanicalMatrixMapper<Rigid3dTypes, Vec1dTypes> >(true)
```
Comment (damienmarchal) [inline modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp]: Same here.

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp [hunk 3]
```diff
@@ -40,18 +40,24 @@ using namespace sofa::defaulttype;
...
 
         ;
 ////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+template class MechanicalMatrixMapper<Rigid3Types, Rigid3Types>;
 template class MechanicalMatrixMapper<Vec3Types, Rigid3Types>;
 template class MechanicalMatrixMapper<Vec3Types, Vec3Types>;
 template class MechanicalMatrixMapper<Vec1Types, Rigid3Types>;
 template class MechanicalMatrixMapper<Vec1Types, Vec1Types>;
+template class MechanicalMatrixMapper<Rigid3dTypes, Vec1dTypes> ;
```
Comment (damienmarchal) [inline modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp]: And here

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.h [hunk 4]
```diff
@@ -164,6 +164,9 @@ class SOFA_GENERAL_ANIMATION_LOOP_API MechanicalMatrixMapper : public MixedInter
 
     Data<helper::vector<std::string>> d_forceFieldList;
     SingleLink < MechanicalMatrixMapper<DataTypes1, DataTypes2>, sofa::simulation::Node , BaseLink::FLAG_STOREPATH > l_nodeToParse;
+    Data <bool> l_stopAtNodeToParse;
```
Comment (damienmarchal) [inline modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.h]: The Sofa convention is to prefix Data field with a d_

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl [hunk 5]
```diff
@@ -57,6 +57,9 @@ MechanicalMatrixMapper<DataTypes1, DataTypes2>::MechanicalMatrixMapper()
     :
       d_forceFieldList(initData(&d_forceFieldList,"forceFieldList","List of ForceField Names to work on (by default will take all)")),
       l_nodeToParse(initLink("nodeToParse","link to the node on which the component will work, from this link the mechanicalState/mass/forceField links will be made")),
+      l_stopAtNodeToParse(initData(&l_stopAtNodeToParse,false,"stopAtNodeToParse","Boolean to choose whether forceFields in children Nodes of NodeToParse should be considered.")),
```
Comment (damienmarchal) [inline modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl]: I look at it and wonder what "parsing" is it talking about. Parsing means 'syntax analysis'. Shouldn't it be "process" the proper term ?

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl [hunk 6]
```diff
@@ -336,7 +338,8 @@ void MechanicalMatrixMapper<DataTypes1, DataTypes2>::addKToMatrix(const Mechanic
 
     sofa::simulation::Node *node = l_nodeToParse.get();
     size_t currentNbInteractionFFs = node->interactionForceField.size();
-
+    msg_info() << "nb m_nbInteractionForceFields :" << m_nbInteractionForceFields;
+    msg_info() << "nb currentNbInteractionFFs :" << currentNbInteractionFFs;
```
Comment (damienmarchal) [inline modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl]: It is better to:
    msg_info() << "nb m_nbInteractionForceFields :" << m_nbInteractionForceFields << msgendl
                        << "nb currentNbInteractionFFs :" << currentNbInteractionFFs;

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl [hunk 7]
```diff
@@ -402,7 +405,8 @@ void MechanicalMatrixMapper<DataTypes1, DataTypes2>::addKToMatrix(const Mechanic
         optimizeAndCopyMappingJacobianToEigenFormat2(J2, J2eig);
         msg_info(this)<<" time set J2eig alone : "<<( (double)timer->getTime() - startTime2)*timeScale<<" ms";
     }
-
+//    msg_info() << "J1:" << J1eig;
```
Comment (damienmarchal) [inline modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl]: Remove.

**Comment by damienmarchal** [comment]
Hi Olivier,

Thanks for your PR. I made some minor comments.

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.h (hunk 4 above)
Comment (olivier-goury) [reply modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.h]: Thank you, that's not a link anyway!

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp (hunk 3 above)
Comment (olivier-goury) [reply modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.cpp]: Thanks, changed!

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl (hunk 5 above)
Comment (olivier-goury) [reply modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl]: You are right. However this Boolean refers to the link "nodeToParse", so I would rather keep the name as is for now. In a future PR, these names should be changed.

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl (hunk 6 above)
Comment (olivier-goury) [reply modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl]: Thank you

File: modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl (hunk 7 above)
Comment (olivier-goury) [reply modules/SofaGeneralAnimationLoop/MechanicalMatrixMapper.inl]: Removed

**Comment by olivier-goury** [comment]
> Hi Olivier,
> 
> Thanks for your PR. I made some minor comments.

Ok I corrected the PR following your comments.

**Comment by guparan** [comment]
[ci-build][with-scene-tests]

**Pull request merged**